pkg/asset/installconfig/aws/validation.go

```diff
@@ -461,24 +530,199 @@ func validateMachineNetworksContainIP(fldPath *field.Path, networks []types.Mach
 	return field.ErrorList{field.Invalid(fldPath, subnetName, fmt.Sprintf("subnet's CIDR range start %s is outside of the specified machine networks", ip))}
 }
 
-func validateDuplicateSubnetZones(fldPath *field.Path, subnets Subnets, idxMap map[string]int, typ string) field.ErrorList {
-	var keys []string
-	for id := range subnets {
-		keys = append(keys, id)
+func validateDuplicateSubnetZones(fldPath *field.Path, subnetDataGroup map[string]subnetData, typ string) field.ErrorList {
+	subnetIDs := make([]string, 0)
+	for id := range subnetDataGroup {
+		subnetIDs = append(subnetIDs, id)
 	}
-	sort.Strings(keys)
+	sort.Strings(subnetIDs)
 
 	allErrs := field.ErrorList{}
 	zones := map[string]string{}
-	for _, id := range keys {
-		subnet := subnets[id]
-		if conflictingSubnet, ok := zones[subnet.Zone.Name]; ok {
-			errMsg := fmt.Sprintf("%s subnet %s is also in zone %s", typ, conflictingSubnet, subnet.Zone.Name)
-			allErrs = append(allErrs, field.Invalid(fldPath.Index(idxMap[id]), id, errMsg))
+	for _, id := range subnetIDs {
+		subnetData := subnetDataGroup[id]
+		if conflictingSubnet, ok := zones[subnetData.Zone.Name]; ok {
+			errMsg := fmt.Sprintf("%s subnet %s is also in zone %s", typ, conflictingSubnet, subnetData.Zone.Name)
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(subnetData.Idx), id, errMsg))
+		} else {
+			zones[subnetData.Zone.Name] = id
+		}
+	}
+	return allErrs
+}
+
+// validateSubnetRoles ensures BYO subnets have valid roles assigned if roles are provided.
+func validateSubnetRoles(fldPath *field.Path, subnetsWithRole map[awstypes.SubnetRoleType][]subnetData, subnetDataGroups subnetDataGroups, publish types.PublishingStrategy) field.ErrorList {
```

Comment:
If the cluster is `publish: External` the bootstrap node would have a public IP address, and therefore should go on a public subnet, I believe. 

I might have missed it, but if it's not there, do you think we should add it?

Reply:
And also updated :D